File: Dynamic Programming/decode-ways.py
# 108ms 6.01%
#28.8MB 5.22%

class Solution:
    def numDecodings(self, s: str) -> int:
        d = []
        if int(s[0]) > 0:
            d.append(1)
        else:
            d.append(0)

        if len(s) >= 2:
            d.append(0)
            if int(s[1]) > 0:
                d[-1] += d[0]
            if 10 <= int(s[:2]) <= 26:
                d[-1] += 1

        for i in range(2, len(s)):
            add = 0
            if 10 <= int(s[i-1:i+1]) <= 26:
                add += d[i-2]
            if int(s[i]) > 0:
                add += d[i-1]
            d.append(add)
        return d[-1]

a = Solution()
l = '130'
print(a.numDecodings(l))

File: Dynamic Programming/decode-ways.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int numDecodings(string s) {
        vector<int> dp(s.size() + 1, 0);
        // this problem has many special cases to consider, like consecutive zeros
        if (s[0] == '0') return 0;
        dp[0] = 1; // two digits are a valid number
        dp[1] = 1;
        for (int i=1; i<s.size(); i++){
            if (s[i] != '0') dp[i+1] += dp[i];
            if (s[i-1] != '0' && stoi(s.substr(i-1, 2)) < 27) dp[i+1] += dp[i-1];
        }
        return dp.back();
    }
};

int main(){
    Solution sol;
    string s = "0";
    cout << sol.numDecodings(s) << endl;
    return 0;
}